rustls/src/conn.rs

```diff
@@ -838,7 +834,8 @@ impl ConnectionCommon {
     }
 }
 
-pub(crate) struct CommonState {
+pub(crate) struct CommonState<Data> {
+    pub(crate) data: Data,
```

Comment:
I'm still not quite sure about this. If we kept an intermediate layer in `ConnectionCommon`, let's just call it `Context` again, with `ConnectionCommon` holding a `Context<Data>` and `Context` holding `data` and `common_state`, we would be able to mutably lend out the `CommonState` to something like the writer, so we could potentially do away with the somewhat awkward `PlaintextSink` API. I'm happy to revert that as a follow-up change, but that would maybe a bit silly. @ctz what do you think?

One reason that would be interesting is that we could implement something like `Deref<CommonState>` and `DerefMut<CommonState>` for `ClientConnection` and `ServerConnection`, which gives us a non-side-specific type which might be useful for downstream users as well as for internal things like reader or writer.

Further thinking on this in the context of #522, at this point my ideal API would probably be to have a public `Connection` type with a type param `Client` or `Server` (and this could also implement `Deref` and `DerefMut`). Those two types would be essentially `ClientConnectionData` and `ServerConnectionData` but we would have them implement some marker trait (`Side`) which has a private supertrait bound. This gives us:

* Precise typing, in that `Connection<T: Side>` can hold methods which apply to both sides, `Connection<Client>` and `Connection<Server>` can hold side-specific methods. Downstream crates could abstract over this as well.
* No more indirection through the `Connection` trait, which seems to cause a decent amount of confusion (in people not being able to find `Connection`'s methods).
* Less code from the `Connection` definition and the fairly trivial impl blocks for it.

I'd prefer to leave most of this stuff for a future PR, but it does inform the design choices here.